File: app/routers/stats.py
"""Ruter Stats?"""

from __future__ import annotations

from datetime import datetime
from textwrap import dedent
from datetime import datetime
from typing import Optional

from fastapi import APIRouter, Depends, Query, Request
from fastapi.responses import HTMLResponse, RedirectResponse
from sqlalchemy.orm import Session

from app.config import settings
from app.db import get_db
from app.models import User, Bot, Destination, Subscription, WebhookEventLog
from app.templating import templates
from app.timezone import TZ

router = APIRouter(tags=["Stats"])


def _check_admin_key(key_from_request: Optional[str]) -> bool:
    admin_key = settings.admin_http_key
    if not admin_key:
        return True
    return (key_from_request or "") == admin_key


def _has_session_admin(request: Request) -> bool:
    user = getattr(request.state, "user", None)
    return bool(user and getattr(user, "is_admin", False))


def _mask_generic(s: Optional[str], keep: int = 3) -> str:
    if not s:
        return "-"
    s = str(s)
    if len(s) <= keep:
        return "•" * len(s)
    return s[:keep] + "…"


def _mask_chat_id(chat_id: Optional[str]) -> str:
    if not chat_id:
        return "-"
    s = str(chat_id)
    # biarkan @username (publik), masker angka
    if s.startswith("@"):
        return s
    # jaga prefix -100 lalu tampilkan 4 digit terakhir
    if s.startswith("-100"):
        return "-100…" + s[-4:]
    if s.startswith("-"):
        return "-…" + s[-4:]
    return _mask_generic(s, keep=3)


def _fmt_dt(value: Optional[datetime]) -> str:
    if not value:
        return "-"
    if value.tzinfo:
        return value.astimezone(TZ).strftime("%Y-%m-%d %H:%M")
    return value.strftime("%Y-%m-%d %H:%M")


@router.get("/stats", response_class=HTMLResponse)
def stats_page(
    request: Request,
    key: Optional[str] = Query(None, alias="key"),
    db: Session = Depends(get_db),
):
    if not _has_session_admin(request):
        if not _check_admin_key(key):
            if key:
                return templates.TemplateResponse(
                    "errors/message.html",
                    {
                        "request": request,
                        "title": "Forbidden",
                        "message": "Invalid admin key.",
                        "page_description": "This dashboard requires a valid admin login or key.",
                        "status_code": 403,
                    },
                    status_code=403,
                )
            login_url = str(request.url_for("auth_login")) + f"?next={request.url.path}"
            return RedirectResponse(login_url, status_code=303)

    total_users = db.query(User).count()
    total_bots = db.query(Bot).count()
    total_dests = db.query(Destination).count()
    total_subs = db.query(Subscription).count()
    total_events = db.query(WebhookEventLog).count()

    # Ringkasan per-user
    users = db.query(User).order_by(User.first_seen_at.asc(), User.id.asc()).all()

    # Subscription terbaru (tanpa hook_id, token, bot_id)
    recent_subs = (
        db.query(Subscription).order_by(Subscription.created_at.desc()).limit(50).all()
    )

    summary = {
        "users": total_users,
        "bots": total_bots,
        "destinations": total_dests,
        "subscriptions": total_subs,
        "events": total_events,
    }

    user_rows = [
        {
            "username": user.username or "-",
            "telegram_masked": _mask_generic(user.telegram_user_id, keep=3),
            "is_admin": user.is_admin,
            "bots": len(user.bots),
            "destinations": len(user.destinations),
            "subscriptions": len(user.subs),
            "first_seen": _fmt_dt(user.first_seen_at),
        }
        for user in users
    ]

    subscription_rows = []
    for sub in recent_subs:
        owner = sub.owner
        dest = sub.destination
        subscription_rows.append(
            {
                "repo": sub.repo,
                "events": sub.events_csv or "*",
                "owner_username": owner.username or "-",
                "owner_masked": _mask_generic(owner.telegram_user_id, keep=3),
                "destination_title": dest.title or "-",
                "destination_masked": _mask_chat_id(dest.chat_id),
                "topic_id": dest.topic_id,
                "created_at": _fmt_dt(sub.created_at),
            }
        )

    recent_events = (
        db.query(WebhookEventLog)
        .order_by(WebhookEventLog.created_at.desc())
        .limit(50)
        .all()
    )

    status_badges = {
        "success": "border-emerald-500/40 bg-emerald-500/10 text-emerald-200",
        "error": "border-red-500/40 bg-red-500/10 text-red-100",
        "ignored": "border-slate-600/40 bg-slate-800/70 text-slate-200",
    }

    event_rows = [
        {
            "created_at": _fmt_dt(log.created_at),
            "event_type": log.event_type or "-",
            "repository": log.repository or "-",
            "status_label": (log.status or "unknown").title(),
            "status_class": status_badges.get(
                (log.status or "").lower(),
                "border-slate-700/50 bg-slate-900/70 text-slate-200",
            ),
            "summary_html": log.summary or "",
            "error": log.error_message,
        }
        for log in recent_events
    ]

    return templates.TemplateResponse(
        "stats/index.html",
        {
            "request": request,
            "page_title": "Stats Dashboard",
            "page_description": "Review users, bots, destinations, and recent GitHub activity across the GitHub → Telegram bridge.",
            "summary": summary,
            "user_rows": user_rows,
            "subscription_rows": subscription_rows,
            "event_rows": event_rows,
        },
    )

File: backend/src/routes/stats.ts
import { Router } from "express";
import { requireAuth } from "../middleware/session";
import { db } from "../db";

const router = Router();

const countUsersStmt = db.prepare("SELECT COUNT(*) as count FROM users");
const countBotsStmt = db.prepare("SELECT COUNT(*) as count FROM bots");
const countDestinationsStmt = db.prepare("SELECT COUNT(*) as count FROM destinations");
const countSubscriptionsStmt = db.prepare("SELECT COUNT(*) as count FROM subscriptions");
const countEventsStmt = db.prepare("SELECT COUNT(*) as count FROM webhook_event_logs");

const listUsersStmt = db.prepare(`
  SELECT
    u.id,
    u.username,
    u.telegram_user_id as telegramUserId,
    u.is_admin as isAdmin,
    u.first_seen_at as firstSeenAt,
    (SELECT COUNT(*) FROM bots b WHERE b.owner_user_id = u.id) AS bots,
    (SELECT COUNT(*) FROM destinations d WHERE d.owner_user_id = u.id) AS destinations,
    (SELECT COUNT(*) FROM subscriptions s WHERE s.owner_user_id = u.id) AS subscriptions
  FROM users u
  ORDER BY u.first_seen_at ASC, u.id ASC
`);

const listSubscriptionsStmt = db.prepare(`
  SELECT
    s.id,
    s.repo,
    s.events_csv as events,
    s.created_at as createdAt,
    u.username,
    u.telegram_user_id as telegramUserId,
    d.title as destinationTitle,
    d.chat_id as destinationChatId,
    d.topic_id as topicId
  FROM subscriptions s
  INNER JOIN users u ON u.id = s.owner_user_id
  INNER JOIN destinations d ON d.id = s.destination_id
  ORDER BY s.created_at DESC
  LIMIT 50
`);

const listEventsStmt = db.prepare(`
  SELECT id, created_at as createdAt, event_type as eventType, repository, status, summary, error_message as error
  FROM webhook_event_logs
  ORDER BY created_at DESC
  LIMIT 50
`);

function maskValue(value: unknown, keep = 3): string {
  if (!value) {
    return "-";
  }
  const str = String(value);
  if (str.startsWith("@")) {
    return str;
  }
  if (str.startsWith("-100") && str.length > 7) {
    return `-100…${str.slice(-4)}`;
  }
  if (str.length <= keep) {
    return "•".repeat(str.length);
  }
  return `${str.slice(0, keep)}…`;
}

function formatDate(value: unknown): string {
  if (!value) {
    return "-";
  }
  const date = new Date(value as string);
  return Number.isNaN(date.getTime()) ? "-" : date.toISOString();
}

router.get("/", requireAuth, (_req, res) => {
  const user = (_req as any).user;
  if (!user?.isAdmin) {
    return res.status(403).json({ error: "forbidden" });
  }

  const summary = {
    users: Number((countUsersStmt.get() as { count: number }).count || 0),
    bots: Number((countBotsStmt.get() as { count: number }).count || 0),
    destinations: Number((countDestinationsStmt.get() as { count: number }).count || 0),
    subscriptions: Number((countSubscriptionsStmt.get() as { count: number }).count || 0),
    events: Number((countEventsStmt.get() as { count: number }).count || 0)
  };

  const users = listUsersStmt.all() as Array<{
    id: number;
    username: string | null;
    telegramUserId: string | null;
    isAdmin: number;
    firstSeenAt: string | null;
    bots: number;
    destinations: number;
    subscriptions: number;
  }>;

  const userRows = users.map((row) => ({
    id: row.id,
    username: row.username || "-",
    telegramMasked: maskValue(row.telegramUserId, 3),
    isAdmin: Boolean(row.isAdmin),
    bots: Number(row.bots || 0),
    destinations: Number(row.destinations || 0),
    subscriptions: Number(row.subscriptions || 0),
    firstSeenAt: formatDate(row.firstSeenAt)
  }));

  const subscriptions = listSubscriptionsStmt.all() as Array<{
    id: number;
    repo: string;
    events: string;
    createdAt: string | null;
    username: string | null;
    telegramUserId: string | null;
    destinationTitle: string | null;
    destinationChatId: string | null;
    topicId: number | null;
  }>;

  const subscriptionRows = subscriptions.map((row) => ({
    id: row.id,
    repo: row.repo,
    events: row.events || "*",
    ownerUsername: row.username || "-",
    ownerMasked: maskValue(row.telegramUserId, 3),
    destinationTitle: row.destinationTitle || "-",
    destinationMasked: maskValue(row.destinationChatId, 3),
    topicId: row.topicId,
    createdAt: formatDate(row.createdAt)
  }));

  const events = listEventsStmt.all() as Array<{
    id: number;
    createdAt: string | null;
    eventType: string | null;
    repository: string | null;
    status: string | null;
    summary: string | null;
    error: string | null;
  }>;

  const eventRows = events.map((row) => ({
    id: row.id,
    createdAt: formatDate(row.createdAt),
    eventType: row.eventType || "unknown",
    repository: row.repository || "-",
    status: row.status || "unknown",
    summary: row.summary || "",
    error: row.error || null
  }));

  return res.json({ summary, users: userRows, subscriptions: subscriptionRows, events: eventRows });
});

export default router;
